src/av_stubs.c

```diff
@@ -0,0 +1,717 @@
+  CAMLparam1(_av);
+  av_t * av = Av_val(_av);
+
+  if( ! av->audio_stream) open_audio_stream(av);
+
+  CAMLreturn(Val_int(av->audio_stream->codec->sample_rate));
+}
+
+CAMLprim value ocaml_av_get_sample_format(value _av) {
+  CAMLparam1(_av);
+  av_t * av = Av_val(_av);
+
+  if( ! av->audio_stream) open_audio_stream(av);
+
+  CAMLreturn(Val_sampleFormat(av->audio_stream->codec->sample_fmt));
+}
+
+static frame_kind decode_packet(av_t * av, stream_t * stream)
+{
+  AVPacket * packet = &av->packet;
+  AVFrame *frame = Frame_val(stream->frame);
+  AVCodecContext * dec = stream->codec;
+  frame_kind frame_kind = undefined;
+  int ret = AVERROR_INVALIDDATA;
+  
+  if(dec->codec_type == AVMEDIA_TYPE_AUDIO ||
+     dec->codec_type == AVMEDIA_TYPE_VIDEO) {
+  
+    caml_release_runtime_system();
+    ret = avcodec_send_packet(dec, packet);
+  
+    if (ret >= 0 || ret == AVERROR(EAGAIN) || ret == AVERROR_EOF) {
+  
+      if (ret >= 0) packet->size = 0;
+  
+      // decode frame
+      ret = avcodec_receive_frame(dec, frame);
+    }
+    caml_acquire_runtime_system();
+  
+    if(dec->codec_type == AVMEDIA_TYPE_AUDIO) {
+      av->audio_stream = stream;
+      frame_kind = audio_frame;
+    }
+    else {
+      av->video_stream = stream;
+      frame_kind = video_frame;
+    }
+  }
+  else if(dec->codec_type == AVMEDIA_TYPE_SUBTITLE) {
+  
+    caml_release_runtime_system();
+    ret = avcodec_decode_subtitle2(dec, stream->subtitle, &stream->got_frame, packet);
+    caml_acquire_runtime_system();
+  
+    if (ret >= 0) packet->size = 0;
+  
+    av->subtitle_stream = stream;
+    frame_kind = subtitle_frame;
+  }
+  
+  if(packet->size <= 0) {
+    av_packet_unref(packet);
+  }
+  
+  if(ret >= 0) {
+    stream->got_frame = 1;
+  }
+  else if (ret == AVERROR_EOF) {
+    stream->got_frame = 0;
+    frame_kind = end_of_file;
+  }
+  else if (ret == AVERROR(EAGAIN)) {
+    frame_kind = undefined;
+  }
+  else {
+    snprintf(error_msg, ERROR_MSG_SIZE, "Error decoding %s frame (%s)", av_get_media_type_string(dec->codec_type), av_err2str(ret));
+    Raise(EXN_FAILURE, error_msg);
+  }
+    
+  return frame_kind;
+}
+
+static value read_stream_frame(av_t * av, stream_t * stream, int media_tag)
+{
+  AVPacket * packet = &av->packet;
+  int stream_index = stream->index;
+  frame_kind frame_kind = undefined;
+    
+  for(; frame_kind == undefined;) {
+    if( ! av->end_of_file) {
+
+      if(packet->size <= 0) {
+	// read frames from the file
+	caml_release_runtime_system();
+	int ret = av_read_frame(av->format, packet);
+	caml_acquire_runtime_system();
+
+	if(ret < 0) {
+	  packet->data = NULL;
+	  packet->size = 0;
+	  av->end_of_file = 1;
+	}
+	else if (packet->stream_index != stream_index) {
+	  av_packet_unref(packet);
+	  continue;
+	}
+      }
+    }
+    frame_kind = decode_packet(av, stream);
+  }
+  
+  if(frame_kind == end_of_file) {
+    return Val_int(END_OF_FILE_TAG);
+  }
+  else {
+    value ans = caml_alloc_small(1, media_tag);
```

Comment:
Ok.
What misled me was that we do this in functions like [value_of_element](https://github.com/savonet/ocaml-gstreamer/blob/master/src/gstreamer_stubs.c#L141).
I will pass ans as parameter.